addons/l10n_fr_certification/models/account.py

```diff
@@ -0,0 +1,153 @@
+
+    def _get_new_hash(self, secure_seq_number):
+        """ Returns the hash to write on journal entries when they get posted"""
+        self.ensure_one()
+        #get the only one exact previous move in the securisation sequence
+        prev_move = self.search([('state', '=', 'posted'),
+                                 ('company_id', '=', self.company_id.id),
+                                 ('l10n_fr_secure_sequence_number', '=', int(secure_seq_number) - 1)])
+        if prev_move and len(prev_move) != 1:
+            raise UserError(
+               _('Error occured when computing the hash. Impossible to get the unique previous posted move'))
+
+        #build and return the hash
+        return self._compute_hash(prev_move.l10n_fr_hash if prev_move else '')
+
+    def _compute_hash(self, previous_hash):
+        """ Computes the hash of the browse_record given as self, based on the hash
+        of the previous record in the company's securisation sequence given as parameter"""
+        self.ensure_one()
+        hash_string = sha1(previous_hash + self.l10n_fr_string_to_hash)
+        return hash_string.hexdigest()
+
+    def _compute_string_to_hash(self):
+        def _getattrstring(obj, field_str):
+            field_value = obj[field_str]
+            if obj._fields[field_str].type == 'many2one':
+                field_value = field_value.id
+            return str(field_value)
+
+        for move in self:
+            values = {}
+            for field in MOVE_FIELDS:
+                values[field] = _getattrstring(move, field)
+
+            for line in move.line_ids:
+                for field in LINE_FIELDS:
+                    values[field] = _getattrstring(line, field)
+            #make the json serialization canonical
+            #  (https://tools.ietf.org/html/draft-staykov-hu-json-canonical-form-00)
+            move.l10n_fr_string_to_hash = dumps(values, sort_keys=True, encoding="utf-8",
+                                                ensure_ascii=True, indent=None,
+                                                separators=(',',':'))
+
+    @api.multi
+    def write(self, vals):
+        has_been_posted = False
+        for move in self:
+            if move.company_id.country_id.code == 'FR':
+                # write the hash and the secure_sequence_number when posting an account.move
+                if vals.get('state') == 'posted':
+                    has_been_posted = True
+
+                # restrict the operation in case we are trying to write a forbidden field
+                if (move.state == "posted" and set(vals).intersection(MOVE_FIELDS)):
+                    raise UserError(ERR_MSG % (self._name, ', '.join(MOVE_FIELDS)))
+                # restrict the operation in case we are trying to overwrite existing hash
+                if (move.l10n_fr_hash and vals.get('l10n_fr_hash')) or (move.l10n_fr_secure_sequence_number and vals.get('l10n_fr_secure_sequence_number')):
```

Comment:
Hi Quentin I think you should only test 
`
if move.l10n_fr_hash or move.l10n_fr_secure_sequence_number:
`
Indeed I I do a write with {'l10n_fr_hash': '', 'l10n_fr_secure_sequence_number': ''} then a write with the new value I will not have any raise error.
What do you think?